python/pyiceberg/table/metadata.py

```diff
@@ -380,19 +383,14 @@ def new_table_metadata(
 class TableMetadataUtil:
     """Helper class for parsing TableMetadata."""
 
-    # Once this has been resolved, we can simplify this: https://github.com/samuelcolvin/pydantic/issues/3846
-    # TableMetadata = Annotated[TableMetadata, Field(alias="format-version", discriminator="format-version")]
+    class _MetadataWrapper(IcebergBaseModel):
+        table_metadata: TableMetadata
 
     @staticmethod
-    def parse_obj(data: Dict[str, Any]) -> TableMetadata:
-        if "format-version" not in data:
-            raise ValidationError(f"Missing format-version in TableMetadata: {data}")
-
-        format_version = data["format-version"]
-
-        if format_version == 1:
-            return TableMetadataV1(**data)
-        elif format_version == 2:
-            return TableMetadataV2(**data)
-        else:
-            raise ValidationError(f"Unknown format version: {format_version}")
+    def parse_raw(data: str) -> TableMetadata:
+        try:
+            d = f'{{"table_metadata": {data}}}'
+            metadata_wrapper = TableMetadataUtil._MetadataWrapper.parse_raw(d)
```

Comment:
The use of the wrapper doesn't make sense to me. Shouldn't 
```suggestion
            metadata_wrapper = TableMetadata.parse_raw(d)
```
work?

Reply:
Btw, I don't like the solution that I found, but it works

Reply:
I see, I also looked into the discriminator to automatically deserialize into the right version. Would be great to get it working, but it is a bit rough it seems.

Reply:
The discriminator works if it is a field in a model. But it does not work per se. We can use it as I did or implement a sort of factory method (as it is right now)

Reply:
I tried to look at what pydantic does and basically it takes the string and then calls json.load. So one of the thing that we can do is:
1. move the json.load call inside this method
2. call the parse_object: now we don't have to concatenate the string since the object is the correct one

Reply:
I'm also bumping Pydantic to 2.0: https://github.com/apache/iceberg/pull/7782 Maybe we can see if that offers something new.